apollo-federation/src/operation/mod.rs

```diff
@@ -200,7 +205,62 @@ pub(crate) struct NormalizedDefer {
     pub operation: Operation,
     pub has_defers: bool,
     pub assigned_defer_labels: IndexSet<String>,
-    pub defer_conditions: IndexMap<String, IndexSet<String>>,
+    pub defer_conditions: IndexMap<Name, IndexSet<String>>,
+}
+
+struct DeferNormalizer {
+    used_labels: HashSet<String>,
+    assigned_labels: IndexSet<String>,
+    conditions: IndexMap<Name, IndexSet<String>>,
+    label_offset: usize,
+}
+
+impl DeferNormalizer {
+    fn new(selection_set: &SelectionSet) -> Result<Self, FederationError> {
+        let mut digest = Self {
+            used_labels: HashSet::default(),
+            label_offset: 0,
+            assigned_labels: IndexSet::default(),
+            conditions: IndexMap::default(),
+        };
+        let mut stack = selection_set
+            .into_iter()
+            .map(|(_, sel)| sel)
+            .collect::<Vec<_>>();
+        while let Some(selection) = stack.pop() {
```

Comment:
should this be recursive instead? we wouldn't need a mutating selection stack

Reply:
I had the exact same thought and the original impl was recursive, but that was what the causing the stack overflows. It is entirely possible that in release mode, those function calls would be optimized away (and there would be tail recursion), but we don't run our tests in release mode.